server/server.js

```diff
@@ -204,20 +230,60 @@ Server.prototype._authorizeMessage = function(socket, message, messageType) {
     isAuthorized = provider.authorize(messageType, message, socket);
   }
 
-  if (!isAuthorized) {
-    logging.warn('#socket.message - auth_invalid', message, socket.id);
-    socket.send({
-      op: 'err',
-      value: 'auth',
-      origin: message
-    });
+  return isAuthorized; 
+};
+
+Server.prototype._limited = function(socket, message, messageType) {
+  var isLimited = false,
+      rateLimiter = this._getRateLimiterForMessageType(messageType);
+
+  if (message.op !== 'subscribe') {
+    return false;
   }
 
-  return isAuthorized; 
+  if (rateLimiter && rateLimiter.isNearLimit(socket.id)) {
+    isLimited = true;
+  }
+
+  return isLimited;
 };
 
-Server.prototype._getMessageType = function(message) {
-  return Type.getByExpression(message.to);
+Server.prototype._updateLimits = function(socket, message, messageType) {
+  var rateLimiter = this._getRateLimiterForMessageType(messageType);
+
+  if (rateLimiter) {
+    switch(message.op) {
+      case 'subscribe': 
+        rateLimiter.add(socket.id, message.to);
+        break;
+      case 'unsubscribe': 
+        rateLimiter.remove(socket.id, message.to);
+        break;
+      default: 
+        logging.debug('#rate_limiting - SKIP update limits', message, socket.id);
+        
+    }
+  }
+};
+
+Server.prototype._getRateLimiterForMessageType = function(messageType) {
+  var rateLimiter;
+
+  if (messageType && messageType.policy && messageType.policy.limit) {
+    rateLimiter = this._rateLimiters[messageType.name];
+
+    if (!rateLimiter) {
+      // TODO: subscribe, as rate limiter operation, should be configurable. 
+      rateLimiter = new RateLimiter(messageType.policy.limit);
+      this._rateLimiters[messageType.name] = rateLimiter;
+    }
+  }
+  
+  return rateLimiter;
+};
+
+Server.prototype._getMessageType = function(messageScope) {
+  return Type.getByExpression(messageScope);
```

Comment:
Actually, Never mind. We only get type twice in the resource lifecycle. So this can be deferred to a later PR.